### DISCUSSION TITLE
cors request

### DISCUSSION THREAD

**Discussion by MoJongleur** [uNetworking/uWebSockets.js#316] [Q&A] [answered]
I could not find any information at all on this topic, is it even possible to setup cors in uWS.js? since I need server permission to accept a request from a specific Url but there is simply no information about this.

I'm trying to just get data from a GET server with a request with a cors header, 
but I get: net :: ERR_EMPTY_RESPONSE

**Comment by mtsewrs** [comment]
Why do think it would not be possible? There is a ton of information on the internet about this, have you tried to google cors?

Seems like you do not know what cors is to begin with.

https://developer.mozilla.org/en-US/docs/Web/HTTP/CORS

**Comment by MoJongleur** [comment]
On stackoverflow could get rid of such comments (go to google), but unfortunately they still remained here. Naturally, I already read this information on mdn, and yes, I'm not a backend developer, maybe I don’t understand something, I look at the analogue of uWS (express) and I see the use of middleware there for such purposes, and try to understand how I can do something like this in uWS.

> Why do think it would not be possible? There is a ton of information on the internet about this, have you tried to google cors?
> 
> Seems like you do not know what cors is to begin with.
> 
> https://developer.mozilla.org/en-US/docs/Web/HTTP/CORS

**Comment by mtsewrs** [comment]
@MoJongleur uws does not have middlewares, so you need to write that yourself. How cors works I suggest you take a look how express is doing it https://github.com/expressjs/cors/blob/master/lib/index.js is a good start

I suggest you learn more about how express works internally (which is not that complicated) and basic backend stuff

**Comment by ghost** [comment]
Isn't this really just setting a header to your response?

**Comment by MoJongleur** [comment, marked as answer]
> Isn't this really just setting a header to your response?

Initially, I thought so, but my poor understanding of the topic did not allow me to understand that this should be done not only in GET but also in OPTION.

Anyway, thanks to everyone for the advice, and sorry for such stupid questions)

**Reply by RJ4100** [reply]
Can you give an example where would you set the header? I'm assuming Access-Control-Allow-Origin would be set in upgradeWSBased handler, correct?

**Comment by ghost** [comment]
Once we have github discussions here we can have a more stackoverflow-ish community where anything goes.

**Comment by lewispham** [comment]
@mtsewrs Setting up cors is not just about adding the access-control-allow-origin header. It also requires serving the preflight requests for non-simple requests. So the OP's issue deserves a better answer than yours.

@MoJongleur In your case of sending a GET request for retrieving data, you'll just need to serve the header `access-control-allow-origin: {client_origin}` with the origin of the client sending the requests. If you want to serve non-simple requests (GraphQL queries for example) then you'll also need to serve the preflight requests.

```javascript
        function setCorsHeaders(response) {
            // You can change the below headers as they're just examples
            response.writeHeader("Access-Control-Allow-Origin", "*");
            response.writeHeader("Access-Control-Allow-Methods", "GET, POST, PUT, DELETE, OPTIONS");
            response.writeHeader("Access-Control-Allow-Headers", "origin, content-type, accept, x-requested-with");
            response.writeHeader("Access-Control-Max-Age", "3600");
        }
        app.options("/graphql", (response) => {
            setCorsHeaders(response);
            response.end();
        });
        app.post("/graphql", (response) => {
            setCorsHeaders(response);
            // handler requests...
        });
```
For more details about preflight requests, you can take a look at this article.
https://developer.mozilla.org/en-US/docs/Glossary/Preflight_request

Basically, uwebsocket.js is a low level library. I actually do not recommend using this lib if you want something to work smoothly right out of the box.

This answer is a bit late but I think it might still help somebody else struggling with cors stuff.

**Reply by e3dio** [reply]
> Basically, uwebsocket.js is a low level library. I actually do not recommend using this lib if you want something to work smoothly right out of the box

Is writing a couple CORS headers really that difficult? Better to understand how CORS works and code it correctly than have server "automatically" handle cross-origin requests for you. Add it to some middleware and you are set

**Reply by ghost** [reply]
I'm also baffled by the conclusion that setting 4 key/values would be "too much work" and "too low level".

**Reply by lewispham** [reply]
First of all, it's not about difficult or easy. It's about mindset. Your statements actually proved my point that features like cors will never be considered. It's actually quite ok to me as I'm expecting uwebsocket.js to have similar level with the core nodejs https.  But as uwebsocket.js is trying to compare performance with `fastify` then I think it's misleading. Developers will think of dropping `fastify` for uwebsocket.js and expecting it to work as smooth as `fastify`. I personally think it's not a good idea because `websocket.js` is a lower level lib than `fastify`.

So why do I think it's low level? To me, as a nodejs user, then anything having same level with nodejs core libs is a low level lib.

Don't get me wrong anyway. I personally like this lib performance and had been using it from its very first days in 2016. I just think developers shouldn't have wrong expectation about it.

**Reply by ghost** [reply]
You don't lose 92% performance by adding four headers. So it's a huge drop down to fastify either way.

**Reply by e3dio** [reply]
> Your statements actually proved my point that features like cors will never be considered

You want a feature that writes a CORS header for you? Writing 2 headers is too much effort? Not even Fastify writes CORS  for you.  There is separate library "fastify-cors" that you need to install and set up that requires more effort than writing the actual 2 CORS headers. You could write a "uws-cors" library but no one would use it because installing a new library just to write 2 headers is ridiculous. Devs these days expect zero code servers with 1 click installs, afraid to put in any effort. Uws.js is 8 times faster than Fastify but it's not worth it because it's "too low level" because you have to write 2 CORS headers. Really ?

**Reply by uasan** [reply]
CORS and performance are not compatible, because CORS makes additional requests for not simple requests, if you think about performance, you should think about a proxy server that will proxy requests to another origin, this will allow the client not to use CORS and reuse only one connection by origin.

**Reply by ghost** [reply]
> Devs these days expect zero code servers with 1 click installs, afraid to put in any effort.

Highly paid developers expect to come to work and play with finger paint these days.

**Reply by lewispham** [reply]
@e3dio I dont need that feature. CORS stuff is just an example for proving that uws.js is a lower level library than fastify. The issue is bigger than just about CORS. `fastify-cors` is a library written by fastify team. It's separated from fastify due to its plugin architecture. Comparing with fastify is not just about the fastify library, it's about comparing with its ecosystem. Does uws.js have the ecosystem like this? https://www.fastify.io/ecosystem/
You can say that there are dedicated libraries for the task. But since uws.js is not compatible with nodejs http core lib, there will be absolutely issues. And resolving those issues will take time, lots of time in some scenarios. So comparing uws.js with fasitfy is completely pointless.

> Devs these days expect zero code servers with 1 click installs, afraid to put in any effort.

What's is so wrong about that expectation? So you want to develop Youtube alike app from scratch with C++ to guarantee performance or something? Conveniency is ways more important than performance in most cases. That's why JavaScript still exists. Performance only matters at large scale level. Fastify is obviously not as fast as uws.js. But it's still fast enough for most apps. So I still do not recommend using uws.js when development time is a huge constraint.

**Reply by lewispham** [reply]
@uasan I'm only using CORS on local for development purpose.

**Reply by e3dio** [reply]
> Does uws.js have the ecosystem like this? https://www.fastify.io/ecosystem

Accepts headers, Auth, Etag Caching, Cookies, Sessions, Websockets, Cors, Forms, Env, JWT, Rate limit, Routes, all things I easily do with a couple lines of code, no need for a separate library for each one. You come from the Zero Code 1 Click Install mindset, that is fine for the finger painters out there, some of us don't need a different library for every 2 headers. I can code more capability faster than you can install a library and read the docs for it, and my server will be 8 times faster :)

**Reply by e3dio** [reply]
You can easily post discussion question for any topic and we can share code, as I already did for your other question

**Comment by rtritto** [comment]
With an ecosystem, we can choose to not reinvent the wheel.

I created a CORS plugin [µWS CORS](https://github.com/uws-ecosystem/uws-cors) (under [µWS Ecosystem](https://github.com/uws-ecosystem)).

Feel free to contribute, to review, to recommend new projects/middlewares/plugins or whatever.

FYI @uNetworkingAB @e3dio @lewispham @porsager
